### PR TITLE
Add mailers

### PR DESCRIPTION
**Pull request by dreamfall**
- Unsuccessful feedback mailer
- Different Buckingham palace invite template for QAEP applications
- couple of other fixes

note, there was an error and I fixed it, but you _need to update notification kinds "unsuccessfull" -> "unsuccessful"_

### CODE REVIEW

File: app/models/email_notification.rb [hunk 1]
```diff
@@ -6,7 +6,7 @@ class EmailNotification < ActiveRecord::Base
 
   belongs_to :settings
 
-  enumerize :kind, in: %w(reminder_to_submit ep_reminder_support_letters winners_notification winners_reminder_to_submit winners_press_release_comments_request unsuccessfull_notification all_unsuccessfull_feedback shortlisted_notifier shortlisted_audit_certificate_reminder not_shortlisted_notifier)
+  enumerize :kind, in: %w(reminder_to_submit ep_reminder_support_letters winners_notification winners_reminder_to_submit winners_press_release_comments_request unsuccessful_notification all_unsuccessful_feedback shortlisted_notifier shortlisted_audit_certificate_reminder not_shortlisted_notifier)
```
Comment (matthewford) [inline app/models/email_notification.rb]: Line is too long. [297/100]

File: app/models/form_answer.rb [hunk 2]
```diff
@@ -94,6 +94,7 @@ def secondary
     scope :for_year, -> (year) { where award_year: year }
     scope :shortlisted_with_no_certificate, -> { where("1 = 0") }
     scope :winners, -> { where("1 = 0") }
+    scope :unsuccessful, -> { where(state: %w(not_recommended not_awarded)) }
```
Comment (matthewford) [inline app/models/form_answer.rb]: Inconsistent indentation detected.

File: app/renderers/mail_renderer.rb [hunk 3]
```diff
@@ -10,7 +10,7 @@ def default_url_options
   end
 
   # this will be removed after all methods are implemented
-  %w(reminder_to_submit ep_reminder_support_letters winners_notification winners_reminder_to_submit unsuccessfull_notification all_unsuccessfull_feedback).each do |method|
+  %w(reminder_to_submit ep_reminder_support_letters winners_notification winners_reminder_to_submit unsuccessful_notification).each do |method|
```
Comment (matthewford) [inline app/renderers/mail_renderer.rb]: Line is too long. [143/100]

File: app/services/notifiers/email_notification_service.rb [hunk 4]
```diff
@@ -18,7 +18,7 @@ def run
   end
 
   # this will be removed after all methods are implemented
-  %w(reminder_to_submit ep_reminder_support_letters winners_reminder_to_submit unsuccessfull_notification all_unsuccessfull_feedback).each do |method|
+  %w(reminder_to_submit ep_reminder_support_letters winners_reminder_to_submit unsuccessful_notification).each do |method|
```
Comment (matthewford) [inline app/services/notifiers/email_notification_service.rb]: Line is too long. [122/100]

File: app/services/notifiers/email_notification_service.rb [hunk 5]
```diff
@@ -42,15 +42,21 @@ def shortlisted_audit_certificate_reminder
...
+  end
+
   def winners_notification
     FormAnswer.winners.each do |form_answer|
-      email = if form_answer.promotion?
-        form_answer.document["nominee_email"]
+      document = form_answer.document
+
+      if form_answer.promotion?
+        Notifiers::Winners::PromotionBuckinghamPalaceInvite.perform_async(document["nominee_email"], form_answer)
```
Comment (matthewford) [inline app/services/notifiers/email_notification_service.rb]: Line is too long. [113/100]

File: app/services/notifiers/email_notification_service.rb [hunk 6]
```diff
@@ -42,15 +42,21 @@ def shortlisted_audit_certificate_reminder
...
       else
-        form_answer.document["head_email"]
+        Notifiers::Winners::BuckinghamPalaceInvite.perform_async(document["head_email"], form_answer)
```
Comment (matthewford) [inline app/services/notifiers/email_notification_service.rb]: Line is too long. [101/100]

File: app/services/notifiers/winners/buckingham_palace_invite.rb [hunk 7]
```diff
@@ -3,9 +3,8 @@ class Notifiers::Winners::BuckinghamPalaceInvite
 
   shoryuken_options queue: "#{Rails.env}_default", auto_delete: true
 
-  def perform(sqs_msg, email)
-    # creating it here doesn't make sense as object won't be valid - missing form_answer
-    invite = PalaceInvite.where(email: email).first_or_create
+  def perform(sqs_msg, email, form_answer_id)
```
Comment (matthewford) [inline app/services/notifiers/winners/buckingham_palace_invite.rb]: Unused method argument - `sqs_msg`. If it's necessary, use `_` or `_sqs_msg` as an argument name to indicate that it won't be used.

File: app/services/notifiers/winners/promotion_buckingham_palace_invite.rb [hunk 8]
```diff
@@ -0,0 +1,10 @@
+class Notifiers::Winners::PromotionBuckinghamPalaceInvite
+  include Shoryuken::Worker
+
+  shoryuken_options queue: "#{Rails.env}_default", auto_delete: true
+
+  def perform(sqs_msg, email, form_answer_id)
```
Comment (matthewford) [inline app/services/notifiers/winners/promotion_buckingham_palace_invite.rb]: Unused method argument - `sqs_msg`. If it's necessary, use `_` or `_sqs_msg` as an argument name to indicate that it won't be used.

File: spec/services/notifiers/email_notification_service_spec.rb [hunk 9]
```diff
@@ -65,16 +63,24 @@
...
+        .with("head@email.com", form_answer)
+      expect(FormAnswer).to receive(:winners) { [form_answer] }
+
+      described_class.run
+
+      expect(current_notification.reload).to be_sent
+    end
+
+    it "triggers current notification" do
+      form_answer = create(:form_answer, :promotion, :submitted, document: { nominee_email: "nominee@email.com" })
```
Comment (matthewford) [inline spec/services/notifiers/email_notification_service_spec.rb]: Line is too long. [114/100]

File: app/models/email_notification.rb [hunk 10]
```diff
@@ -6,7 +6,18 @@ class EmailNotification < ActiveRecord::Base
 
   belongs_to :settings
 
-  enumerize :kind, in: %w(reminder_to_submit ep_reminder_support_letters winners_notification winners_reminder_to_submit winners_press_release_comments_request unsuccessfull_notification all_unsuccessfull_feedback shortlisted_notifier shortlisted_audit_certificate_reminder not_shortlisted_notifier)
+  enumerize :kind, in: [
+                         :reminder_to_submit,
```
Comment (matthewford) [inline app/models/email_notification.rb]: Use 2 spaces for indentation in an array, relative to the start of the line where the left bracket is.

File: app/models/email_notification.rb [hunk 11]
```diff
@@ -6,7 +6,18 @@ class EmailNotification < ActiveRecord::Base
...
+                         :ep_reminder_support_letters,
+                         :winners_notification,
+                         :winners_reminder_to_submit,
+                         :winners_press_release_comments_request,
+                         :unsuccessful_notification,
+                         :all_unsuccessful_feedback,
+                         :shortlisted_notifier,
+                         :shortlisted_audit_certificate_reminder,
+                         :not_shortlisted_notifier
+                       ]
```
Comment (matthewford) [inline app/models/email_notification.rb]: Indent the right bracket the same as the start of the line where the left bracket is.

File: app/models/form_answer.rb (hunk 2 above)
Comment (matthewford) [inline app/models/form_answer.rb]: Inconsistent indentation detected.

File: app/services/notifiers/email_notification_service.rb (hunk 4 above)
Comment (matthewford) [inline app/services/notifiers/email_notification_service.rb]: Line is too long. [122/100]

File: app/services/notifiers/email_notification_service.rb (hunk 5 above)
Comment (matthewford) [inline app/services/notifiers/email_notification_service.rb]: Line is too long. [113/100]

File: app/services/notifiers/email_notification_service.rb (hunk 6 above)
Comment (matthewford) [inline app/services/notifiers/email_notification_service.rb]: Line is too long. [101/100]

File: app/services/notifiers/email_notification_service.rb [hunk 12]
```diff
@@ -63,4 +71,10 @@ def winners_press_release_comments_request
       end
     end
   end
+
+  private
+
+  def current_award_year
```
Comment (jaleszek) [inline app/services/notifiers/email_notification_service.rb]: Maybe add new issue ? I'm a bit afraid that we stuck with priorities of all these todo comments. What do you think ?

File: spec/mailers/users/audit_certificate_request_mailer_spec.rb [hunk 13]
```diff
@@ -14,7 +14,6 @@
   }
 
   before do
-    allow_any_instance_of(FormAnswer).to receive(:eligible?) { true }
```
Comment (jaleszek) [inline spec/mailers/users/audit_certificate_request_mailer_spec.rb]: :+1:

File: app/models/form_answer.rb (hunk 2 above)
Comment (jaleszek) [inline app/models/form_answer.rb]: @dreamfall so unsuccessful app is only in these to states ? What about withdrawn, not_eligible, not_submitted.

File: app/models/email_notification.rb (hunk 10 above)
Comment (matthewford) [inline app/models/email_notification.rb]: Use 2 spaces for indentation in an array, relative to the start of the line where the left bracket is.

**Pull request merged**